provider/factory/factory_test.go

```diff
@@ -0,0 +1,54 @@
+package factory_test
+
+import (
+	"errors"
+	"testing"
+
+	"github.com/cycloidio/inframap/errcode"
+	"github.com/cycloidio/inframap/provider"
+	"github.com/cycloidio/inframap/provider/factory"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestGetProviderandResource(t *testing.T) {
+	tests := []struct {
+		Name      string
+		Input     string
+		EProvider provider.Type
+		EResource string
+		EError    error
+	}{
+		{
+			Name:      "Success",
+			Input:     "aws_lb.front",
+			EProvider: provider.AWS,
+			EResource: "aws_lb",
+		},
+		{
+			Name:      "SuccessOnlyResource",
+			Input:     "aws_lb",
+			EProvider: provider.AWS,
+			EResource: "aws_lb",
+		},
+		{
+			Name:   "ErrProviderNotFound-InvalidProvider",
+			Input:  "pepe_a.front",
+			EError: errcode.ErrProviderNotFound,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.Name, func(t *testing.T) {
+			p, rs, err := factory.GetProviderAndResource(tt.Input)
+
+			if tt.EError == nil {
+				require.NoError(t, err)
+				assert.Equal(t, tt.EResource, rs)
+				assert.Equal(t, tt.EProvider, p.Type())
+			} else {
+				assert.True(t, errors.Is(err, tt.EError))
+			}
```

Comment:
Hmm, couldn't we do the asserts every loop?
I think we don't need the condition and `require.NoError`, just the asserts

Reply:
I think it lets the code flow much easy to read and overall this is what we would normally do:
* If error expected check error
* If no error expected check the results and make sure no error is returned.

Also if unified I would also need a condition to check if `p` is nil (in case of error) as the `p.Type()` would panic then.